src/dotnet/Azure.ClientSdk.Analyzers/Azure.ClientSdk.Analyzers/AddConfigureAwaitAnalyzer.cs

```diff
@@ -0,0 +1,131 @@
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
+            ImmutableArray.Create(Descriptors.AZC0012, Descriptors.AZC0013);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.RegisterCompilationStartAction(CompilationStart);
+        }
+
+        private void CompilationStart(CompilationStartAnalysisContext context) 
+        {
+            _asyncUtilities = new AsyncAnalyzerUtilities(context.Compilation);
+            context.RegisterSyntaxNodeAction(AnalyzeAwaitExpression, SyntaxKind.AwaitExpression);
+            context.RegisterSyntaxNodeAction(AnalyzeUsingExpression, SyntaxKind.UsingStatement);
+            context.RegisterSyntaxNodeAction(AnalyzeForEachExpression, SyntaxKind.ForEachStatement);
+            context.RegisterSyntaxNodeAction(AnalyzeConfigureAwaitTrue, SyntaxKind.InvocationExpression);
+        }
+
+        private void AnalyzeAwaitExpression(SyntaxNodeAnalysisContext context)
+        {
+            var invocation = (AwaitExpressionSyntax)context.Node;
+            var operation = context.SemanticModel.GetOperation(invocation, context.CancellationToken);
+            if (!(operation is IAwaitOperation awaitOperation)) {
+                return;
+            }
+
+            if (_asyncUtilities.IsTaskType(awaitOperation.Operation.Type)) 
+            {
+                ReportConfigureAwaitDiagnostic(context, awaitOperation);
+            }
+        }
+
+        private void AnalyzeUsingExpression(SyntaxNodeAnalysisContext context) 
+        {
+            if (!(context.SemanticModel.GetOperation(context.Node, context.CancellationToken) is IUsingOperation usingOperation))
+            {
+                return;
+            }
+
+            if (!usingOperation.IsAsynchronous)
+            {
+                return;
+            }
+
+            var resources = usingOperation.Resources;
+            if (_asyncUtilities.IsAsyncDisposableType(resources.Type))
+            {
+                ReportConfigureAwaitDiagnostic(context, resources);
+            }
+        }
+
+        private void AnalyzeForEachExpression(SyntaxNodeAnalysisContext context) 
+        {
+            if (!(context.SemanticModel.GetOperation(context.Node, context.CancellationToken) is IForEachLoopOperation forEachOperation))
+            {
+                return;
+            }
+
+            if (!forEachOperation.IsAsynchronous)
+            {
+                return;
+            }
+
+            var collectionType = forEachOperation.Collection.Type;
+            if (_asyncUtilities.IsAsyncEnumerableType(collectionType)) 
+            {
+                ReportConfigureAwaitDiagnostic(context, forEachOperation.Collection);
+            }
+        }
+
+        private void AnalyzeConfigureAwaitTrue(SyntaxNodeAnalysisContext context)
+        {
+            if (!_asyncUtilities.IsConfigureAwait(context.Node))
```

Comment:
Does this check that the configure await is set to false?

Would this cover cases where we have nested operations? I assume it does since the analyzer will be invoked for each await operation -- but it would be nice to add a test if we don't have one already.

Reply:
> Would this cover cases where we have nested operations?

Can you give an example?

Reply:
I was thinking of this: 

```csharp
// defs
async Task FooAsync(bool foo) {  }
async Task<bool> getFoo() { }

// call
await FooAsync( await getFoo() );

// with the configure await
await FooAsync( (await getFoo()).ConfigureAwait(false) ).ConfigureAwait(false);

```

Reply:
Added this as a test